src/backend/distributed/executor/multi_executor.c

```diff
@@ -24,273 +24,442 @@
-			 * planned statement is now replaced we can't call GetMultiPlan() in
-			 * the later hooks, so we set a flag marking this as a distributed
-			 * statement running on the master. That e.g. allows us to drop the
-			 * temp table later.
-			 *
-			 * We copy the original statement's queryId, to allow
-			 * pg_stat_statements and similar extension to associate the
-			 * statement with the toplevel statement.
-			 */
-			masterSelectPlan->queryId = queryDesc->plannedstmt->queryId;
-			queryDesc->plannedstmt = masterSelectPlan;
-
-			eflags |= EXEC_FLAG_CITUS_MASTER_SELECT;
+	multiPlan = scanState->multiPlan;
+	workerJob = multiPlan->workerJob;
+	taskList = workerJob->taskList;
+
+	/* check if this is a single shard query */
+	if (list_length(taskList) == 1)
+	{
+		bool isModificationQuery = IsModifyMultiPlan(multiPlan);
+		if (isModificationQuery)
+		{
+			scanState->customScanState.methods = &RouterSingleModifyCustomExecMethods;
+		}
+		else
+		{
+			scanState->customScanState.methods = &RouterSelectCustomExecMethods;
 		}
 	}
-
-	/* if the execution is not done for router executor, drop into standard executor */
-	if (queryDesc->estate == NULL ||
-		!(queryDesc->estate->es_top_eflags & EXEC_FLAG_CITUS_ROUTER_EXECUTOR))
+	else
 	{
-		standard_ExecutorStart(queryDesc, eflags);
+		scanState->customScanState.methods = &RouterMultiModifyCustomExecMethods;
 	}
+
+	return (Node *) scanState;
 }
 
 
 /*
- * CopyQueryResults executes the commands that copy query results into a
- * temporary table.
+ * InvalidCreateScan is only called on an invalid case which we would like to
+ * error out. This is the case when a plan is not ready for execution because
+ * CreateDistributedPlan() couldn't find a plan due to unresolved prepared
+ * statement parameters, but didn't error out, because we expect custom plans
+ * to come to our rescue. But sql (not plpgsql) functions unfortunately don't
+ * go through a codepath supporting custom plans. We call VerifyMultiPlanValidity()
+ * to do this check and provide a meaningfull error message.
  */
-static void
-CopyQueryResults(List *masterCopyStmtList)
+Node *
+InvalidCreateScan(CustomScan *scan)
 {
-	ListCell *masterCopyStmtCell = NULL;
+	CitusScanState *scanState = palloc0(sizeof(CitusScanState));
 
-	/* now copy data from all the remote nodes into temp table */
-	foreach(masterCopyStmtCell, masterCopyStmtList)
-	{
-		Node *masterCopyStmt = (Node *) lfirst(masterCopyStmtCell);
+	scanState->executorType = MULTI_EXECUTOR_INVALID_FIRST;
+	scanState->customScanState.ss.ps.type = T_CustomScanState;
+	scanState->multiPlan = GetMultiPlan(scan);
 
-		Assert(IsA(masterCopyStmt, CopyStmt));
+	Assert(IsA(scanState, CustomScanState));
 
-		ProcessUtility(masterCopyStmt,
-					   "(copy job)",
-					   PROCESS_UTILITY_QUERY,
-					   NULL,
-					   None_Receiver,
-					   NULL);
-	}
+	/* ensure plan is executable */
+	VerifyMultiPlanValidity(scanState->multiPlan);
```

Comment:
@anarazel Hmm, I didn't quite get that. Don't we use the same logic to error out cases with parameterized sql functions? What is the different now?